File: 498_diagonal_traverse.java
class Solution {
    public int[] findDiagonalOrder(int[][] matrix) {
        if (matrix == null || matrix.length == 0) return new int[0];

        int row = matrix.length, col = matrix[0].length, total = row * col;
        int i = 0, j = 0, count = 1;
        boolean isUpper = true;
        int[] result = new int[total];
        while (count <= total) {
            result[count - 1] = matrix[i][j];
            if (isUpper) {
                if (i - 1 < 0 && j + 1 < col) {   // move to right ele
                    j += 1;
                    isUpper = false;
                } else if (i - 1 < 0 && j + 1 == col) {  // move to down ele
                    i += 1;
                    isUpper = false;
                } else if (i - 1 >= 0 && j + 1 >= col) {  // move to down ele
                    i += 1;
                    isUpper = false;
                } else {
                    i -= 1;
                    j += 1;
                }
            } else {
                if (i + 1 < row && j - 1 < 0) { // move to down ele
                    i += 1;
                    isUpper = true;
                } else if (i + 1 == row && j - 1 < 0) {  // move to right ele
                    j += 1;
                    isUpper = true;
                } else if (i + 1 >= row && j - 1 >= 0) {  // move to right ele
                    j += 1;
                    isUpper = true;
                } else {
                    i += 1;
                    j -= 1;
                }
            }
            count ++;
        }
        return result;
    }
}

File: python/498_diagonal_traverse.py
class Solution:
    # Time Complexity: O(n)
    # Space Complexity: O(n)
    def findDiagonalOrder(self, matrix: List[List[int]]) -> List[int]:
        if not matrix or not matrix[0]:
            return []

        M, N = len(matrix), len(matrix[0])
        total = M * N
        isUp = True
        res = []
        i = j = count = 0
        while count < total:
            res.append(matrix[i][j])
            if isUp:
                # at left up
                if i - 1 < 0 and j + 1 < N:
                    j += 1
                    isUp = False
                # just actually at right up corner
                elif i - 1 == -1 and j + 1 == N:
                    i += 1
                    isUp = False
                # at right down
                elif i - 1 >= 0 and j + 1 >= N:
                    i += 1
                    isUp = False
                # while increasing
                else:
                    i -= 1
                    j += 1
            else:
                if i + 1 < M and j - 1 < 0:
                    i += 1
                    isUp = True
                elif i + 1 == M and j - 1 == -1:
                    j += 1
                    isUp = True
                elif i + 1 >= M and j - 1 >= 0:
                    j += 1
                    isUp = True
                else:
                    i += 1
                    j -= 1
            count += 1
        return res





